File: src/lib/keybindings.ts
import { getSetting, setSetting } from "./settings";
import { invoke } from "@tauri-apps/api/core";

export type ActionContext = "browser" | "app";

export interface ActionDef {
  id: string;
  label: string;
  context: ActionContext;
  defaultKey: string;
}

export const ACTION_DEFINITIONS: ActionDef[] = [
  // Browser vimium actions
  { id: "scroll_down",      label: "Scroll down",         context: "browser", defaultKey: "j"  },
  { id: "scroll_up",        label: "Scroll up",           context: "browser", defaultKey: "k"  },
  { id: "scroll_half_down", label: "Scroll ½ down",       context: "browser", defaultKey: "d"  },
  { id: "scroll_half_up",   label: "Scroll ½ up",         context: "browser", defaultKey: "u"  },
  { id: "scroll_to_top",    label: "Scroll to top",       context: "browser", defaultKey: "gg" },
  { id: "scroll_to_bottom", label: "Scroll to bottom",    context: "browser", defaultKey: "G"  },
  { id: "history_back",     label: "History back",        context: "browser", defaultKey: "H"  },
  { id: "history_forward",  label: "History forward",     context: "browser", defaultKey: "L"  },
  { id: "reload",           label: "Reload page",         context: "browser", defaultKey: "r"  },
  { id: "hint_mode",        label: "Hint mode (click)",   context: "browser", defaultKey: "f"  },

  // App-level action slots (unimplemented — reserved for future use)
  { id: "open_settings",    label: "Open settings",       context: "app",     defaultKey: "Ctrl+," },
  { id: "focus_chat",       label: "Focus chat input",    context: "app",     defaultKey: "Ctrl+l" },
];

/** Overrides map: action id → user-set key (only entries that differ from default). */
export type Overrides = Record<string, string>;

/** Returns the effective key for every action, merging defaults with overrides. */
export function resolveBindings(overrides: Overrides): Record<string, string> {
  const result: Record<string, string> = {};
  for (const def of ACTION_DEFINITIONS) {
    result[def.id] = overrides[def.id] ?? def.defaultKey;
  }
  return result;
}

/** Loads stored overrides from SQLite. Returns {} if none saved yet. */
export async function loadOverrides(): Promise<Overrides> {
  const raw = await getSetting("keybindings");
  if (!raw) return {};
  try {
    return JSON.parse(raw) as Overrides;
  } catch {
    return {};
  }
}

/**
 * Saves overrides to SQLite and syncs Rust managed state so the next
 * browser page load picks up the new bindings.
 */
export async function saveOverrides(overrides: Overrides): Promise<void> {
  await setSetting("keybindings", JSON.stringify(overrides));
  await invoke("sync_keybindings", { overrides });
}

/**
 * Checks whether `proposedKey` conflicts with any existing binding in the
 * same context as `actionId`.
 *
 * Returns a human-readable conflict description, or null if no conflict.
 * Blocks both direct duplicates and prefix relationships.
 */
export function detectConflict(
  actionId: string,
  proposedKey: string,
  effectiveMap: Record<string, string>
): string | null {
  const action = ACTION_DEFINITIONS.find((a) => a.id === actionId);
  if (!action) return null;

  for (const def of ACTION_DEFINITIONS) {
    if (def.id === actionId) continue;
    if (def.context !== action.context) continue;

    const existing = effectiveMap[def.id];
    if (!existing) continue;

    if (existing === proposedKey) {
      return `Already bound to "${def.label}"`;
    }
    if (existing.startsWith(proposedKey)) {
      return `Prefix of "${def.label}" (${existing})`;
    }
    if (proposedKey.startsWith(existing)) {
      return `"${def.label}" (${existing}) is a prefix of this key`;
    }
  }
  return null;
}

File: src-tauri/src/commands/keybindings.rs
use std::collections::HashMap;
use std::sync::Mutex;
use tauri::State;

pub const DEFAULT_BINDINGS: &[(&str, &str)] = &[
    ("scroll_down",      "j"),
    ("scroll_up",        "k"),
    ("scroll_half_down", "d"),
    ("scroll_half_up",   "u"),
    ("scroll_to_top",    "gg"),
    ("scroll_to_bottom", "G"),
    ("history_back",     "H"),
    ("history_forward",  "L"),
    ("reload",           "r"),
    ("hint_mode",        "f"),
];

// App-context actions and their defaults. Key format matches keyEventToString in App.tsx.
pub const DEFAULT_APP_BINDINGS: &[(&str, &str)] = &[
    ("open_settings",   "Ctrl+,"),
    ("focus_chat",      "Alt+n"),
    ("focus_workspace", "Alt+m"),
    ("tab_prev",        "Alt+j"),
    ("tab_next",        "Alt+k"),
    ("tab_first",       "Alt+1"),
    ("tab_last",        "Alt+0"),
];

/// Holds user overrides: action_id → custom key.
/// Absent means "use default".
pub struct KeybindingState(pub Mutex<HashMap<String, String>>);

impl Default for KeybindingState {
    fn default() -> Self {
        KeybindingState(Mutex::new(HashMap::new()))
    }
}

/// Builds the JS snippet that sets window.__bindings and window.__app_bindings,
/// merging defaults with overrides. Key format for app bindings matches keyEventToString in App.tsx.
pub fn bindings_js(overrides: &HashMap<String, String>) -> String {
    let browser_entries: Vec<String> = DEFAULT_BINDINGS
        .iter()
        .map(|(id, default_key)| {
            let key = overrides.get(*id).map(String::as_str).unwrap_or(default_key);
            format!("\"{}\":\"{}\"", id, key.replace('"', "\\\""))
        })
        .collect();
    let app_entries: Vec<String> = DEFAULT_APP_BINDINGS
        .iter()
        .map(|(id, default_key)| {
            let key = overrides.get(*id).map(String::as_str).unwrap_or(default_key);
            format!("\"{}\":\"{}\"", id, key.replace('"', "\\\""))
        })
        .collect();
    format!(
        "window.__bindings={{{}}};window.__app_bindings={{{}}};",
        browser_entries.join(","),
        app_entries.join(",")
    )
}

/// Called from the frontend after the user saves bindings.
/// Updates managed state so the next page load picks up the new bindings.
#[tauri::command]
pub fn sync_keybindings(
    state: State<'_, KeybindingState>,
    overrides: HashMap<String, String>,
) -> Result<(), String> {
    *state.0.lock().map_err(|e| e.to_string())? = overrides;
    Ok(())
}
